esp-wifi/src/lib.rs

```diff
@@ -361,7 +385,7 @@ impl EspWifiTimerSource for TimeBase {
 /// ```
 pub fn init<'d, T: EspWifiTimerSource>(
     timer: impl Peripheral<P = T> + 'd,
-    _rng: hal::rng::Rng,
+    _rng: impl EspWifiRngSource,
```

Comment:
It could be, but the problem is that we don't want to completely own the Rng or Trng here, users will still want to use it in their own applications see: https://github.com/esp-rs/esp-hal/issues/2372

Reply:
And how does this (a trait that requires passing by value) help in not completely owning the rng?

Reply:
`Rng` driver is `clone`, `Trng` might be clone soon. `RNG` (the raw peripheral) will take exclusive access whether its via `&mut Self` or `Self` and not allow any other part of the system to use `Rng`.

Reply:
I think you misunderstand what I want. I'm not proposing requiring using the raw peripheral. I'm proposing accepting drivers that work with the peripheral. Making `Rng` and `Trng` implement `Peripheral` allows us to pass both of those, or a mutable reference to them (or just the RNG peripheral directly) and esp-wifi wouldn't know the difference.

I.e. I'm suggesting that instead of a custom trait we have:

```rust
impl Peripheral for Rng {
    type P = RNG;
...
}


impl Peripheral for Trng {
    type P = RNG;
...
}
```

Reply:
Ah I see, yeah I misunderstood initially. I like this much more, I'll do a follow up PR now.

Reply:
Ah, the reason _not_ to do this is that a user can then pass in `peripherals.RNG` by itself, which won't have actually initialized the RNG, which is an assumption we're making in esp-wifi.

Reply:
But what initialization does Rng need? It's constructor is a nop :)

Reply:
Only because the esp-idf bootloader does the init for us, which reminds me I need to file an issue for that. We can't depend on this being the case, especially if we move to mcuboot, so eventually this constructor will do something.

Reply:
https://github.com/esp-rs/esp-hal/issues/2460

Reply:
In that case we may still go this way, just using `Peripheral<P = Rng>`. But I'll not force the issue.

Reply:
With `Trng` also impl `Peripheral<P = Rng>`? That could work. Just wondering if there a case where someone might need to write `Peripheral<P = Trng>`, because they wouldn't be able to in this case.

Reply:
> `Peripheral<P = Trng>`

That is an interesting question, drivers absolutely requiring true randomness. I'm not sure, but we're not supporting it right now, either.

Reply:
The "funny" thing is that RNG will give true random numbers when the RF peripherals are enabled (but we still need #2443)